### DISCUSSION TITLE
Need help with ljspeech\xtts_v1\train_gpt_xtts.py - Fine tuning a single voice - configuration and warnings and errors

### DISCUSSION THREAD

**Discussion by FurkanGozukara** [coqui-ai/TTS#3167] [General Q&A]
I am trying to make training work official raw dataset of LJSpeech-1.1. after making it work hopefully i will train on my own

My aim is fine tuning a single voice to get eleven labs level quality

Then hopefully I will release this as a tutorial on my AI channel SECourses

https://www.youtube.com/SECourses

So i did setup everything.  I have RTX 3090 TI and 13900 K CPU

I need help with configuration and also output warnings and errors

Platform  : 
Windows 10
Python 3.10.11
VENV
RTX 3090 TI
Pytorch 2.0.1 cuda 118

**Here the config I am running**

```
# Logging parameters
RUN_NAME = "GPT_XTTS_LJSpeech_FT"
PROJECT_NAME = "XTTS_trainer"
DASHBOARD_LOGGER = "tensorboard"
LOGGER_URI = None

# Set here the path that the checkpoints will be saved. Default: ./run/training/
OUT_PATH = r"F:\coqui_test\train"

# Training Parameters
OPTIMIZER_WD_ONLY_ON_WEIGHTS = True  # for multi-gpu training please make it False
START_WITH_EVAL = True  # if True it will star with evaluation
BATCH_SIZE = 3  # set here the batch size
GRAD_ACUMM_STEPS = 84  # set here the grad accumulation steps
# Note: we recommend that BATCH_SIZE * GRAD_ACUMM_STEPS need to be at least 252 for more efficient training. You can increase/decrease BATCH_SIZE but then set GRAD_ACUMM_STEPS accordingly.

# Define here the dataset that you want to use for the fine-tuning on.
config_dataset = BaseDatasetConfig(
    formatter="ljspeech",
    dataset_name="ljspeech",
    path=r"F:\coqui_test\LJSpeech-1.1\wavs",
    meta_file_train=r"F:\coqui_test\LJSpeech-1.1\metadata.csv",
    language="en",
)

# Add here the configs of the datasets
DATASETS_CONFIG_LIST = [config_dataset]

# Define the path where XTTS v1.1.1 files will be downloaded
CHECKPOINTS_OUT_PATH = r"F:\coqui_test\check_point_download"
os.makedirs(CHECKPOINTS_OUT_PATH, exist_ok=True)


# DVAE files
DVAE_CHECKPOINT_LINK = "https://coqui.gateway.scarf.sh/hf-coqui/XTTS-v1/v1.1.1/dvae.pth"
MEL_NORM_LINK = "https://coqui.gateway.scarf.sh/hf-coqui/XTTS-v1/v1.1.1/mel_stats.pth"

# Set the path to the downloaded files
DVAE_CHECKPOINT = os.path.join(CHECKPOINTS_OUT_PATH, DVAE_CHECKPOINT_LINK.split("/")[-1])
MEL_NORM_FILE = os.path.join(CHECKPOINTS_OUT_PATH, MEL_NORM_LINK.split("/")[-1])

# download DVAE files if needed
if not os.path.isfile(DVAE_CHECKPOINT) or not os.path.isfile(MEL_NORM_FILE):
    print(" > Downloading DVAE files!")
    ModelManager._download_model_files([MEL_NORM_LINK, DVAE_CHECKPOINT_LINK], CHECKPOINTS_OUT_PATH, progress_bar=True)


# Download XTTS v1.1 checkpoint if needed
TOKENIZER_FILE_LINK = "https://coqui.gateway.scarf.sh/hf-coqui/XTTS-v1/v1.1.1/vocab.json"
XTTS_CHECKPOINT_LINK = "https://coqui.gateway.scarf.sh/hf-coqui/XTTS-v1/v1.1.1/model.pth"

# XTTS transfer learning parameters: You we need to provide the paths of XTTS model checkpoint that you want to do the fine tuning.
TOKENIZER_FILE = os.path.join(CHECKPOINTS_OUT_PATH, TOKENIZER_FILE_LINK.split("/")[-1]) # vocab.json file
XTTS_CHECKPOINT = os.path.join(CHECKPOINTS_OUT_PATH, XTTS_CHECKPOINT_LINK.split("/")[-1]) # model.pth file

# download XTTS v1.1 files if needed
if not os.path.isfile(TOKENIZER_FILE) or not os.path.isfile(XTTS_CHECKPOINT):
    print(" > Downloading XTTS v1.1 files!")
    ModelManager._download_model_files([TOKENIZER_FILE_LINK, XTTS_CHECKPOINT_LINK], CHECKPOINTS_OUT_PATH, progress_bar=True)


# Training sentences generations
SPEAKER_REFERENCE = (
    "./tests/data/ljspeech/wavs/LJ001-0002.wav"  # speaker reference to be used in training test sentences
)
LANGUAGE = config_dataset.language


def main():
    # init args and config
    model_args = GPTArgs(
        max_conditioning_length=132300,  # 6 secs
        min_conditioning_length=66150,  # 3 secs
        debug_loading_failures=False,
        max_wav_length=255995,  # ~11.6 seconds
        max_text_length=200,
        mel_norm_file=MEL_NORM_FILE,
        dvae_checkpoint=DVAE_CHECKPOINT,
        # tokenizer_file="/raid/datasets/xtts_models/vocab.json", # vocab path of the model that you want to fine-tune
        # xtts_checkpoint="https://huggingface.co/coqui/XTTS-v1/resolve/hifigan/model.pth",
        xtts_checkpoint=XTTS_CHECKPOINT,  # checkpoint path of the model that you want to fine-tune
        tokenizer_file=TOKENIZER_FILE,
        gpt_num_audio_tokens=8194,
        gpt_start_audio_token=8192,
        gpt_stop_audio_token=8193,
        use_ne_hifigan=True, # if it is true it will keep the non-enhanced keys on the output checkpoint
    )
    # define audio config
    audio_config = XttsAudioConfig(
        sample_rate=22050, dvae_sample_rate=22050, diffusion_sample_rate=24000, output_sample_rate=24000
    )
    # training parameters config
    config = GPTTrainerConfig(
        output_path=OUT_PATH,
        model_args=model_args,
        run_name=RUN_NAME,
        project_name=PROJECT_NAME,
        run_description="""
            GPT XTTS training
            """,
        dashboard_logger=DASHBOARD_LOGGER,
        logger_uri=LOGGER_URI,
        audio=audio_config,
        batch_size=BATCH_SIZE,
        batch_group_size=48,
        eval_batch_size=BATCH_SIZE,
        num_loader_workers=8,
		num_eval_loader_workers=4,
        eval_split_max_size=256,
        print_step=10,
        plot_step=100,
		mixed_precision=True,
		text_cleaner="english_cleaners",
		use_phonemes=True,
		phoneme_language="en-us",
        log_model_step=1,
        save_step=10000,
        save_n_checkpoints=100,
        save_checkpoints=True,
        # target_loss="loss",
        print_eval=True,
        # Optimizer values like tortoise, pytorch implementation with modifications to not apply WD to non-weight parameters.
        optimizer="AdamW",
        optimizer_wd_only_on_weights=OPTIMIZER_WD_ONLY_ON_WEIGHTS,
        optimizer_params={"betas": [0.9, 0.96], "eps": 1e-8, "weight_decay": 1e-2},
        lr=5e-06,  # learning rate
        lr_scheduler="MultiStepLR",
        # it was adjusted accordly for the new step scheme
        lr_scheduler_params={"milestones": [50000 * 18, 150000 * 18, 300000 * 18], "gamma": 0.5, "last_epoch": -1},
        test_sentences=[
            {
                "text": "It took me quite a long time to develop a voice, and now that I have it I'm not going to be silent.",
                "speaker_wav": SPEAKER_REFERENCE,
                "language": LANGUAGE,
            },
            {
                "text": "This cake is great. It's so delicious and moist.",
                "speaker_wav": SPEAKER_REFERENCE,
                "language": LANGUAGE,
            },
        ],
    )

    # init the model from config
    model = GPTTrainer.init_from_config(config)

    # load training samples
    train_samples, eval_samples = load_tts_samples(
        DATASETS_CONFIG_LIST,
        eval_split=True,
        eval_split_max_size=config.eval_split_max_size,
        eval_split_size=config.eval_split_size,
    )

    # init the trainer and 🚀
    trainer = Trainer(
        TrainerArgs(
            restore_path=None,  # xtts checkpoint is restored via xtts_checkpoint key so no need of restore it using Trainer restore_path parameter
            skip_train_epoch=False,
            start_with_eval=START_WITH_EVAL,
            grad_accum_steps=GRAD_ACUMM_STEPS,
        ),
        config,
        output_path=OUT_PATH,
        model=model,
        train_samples=train_samples,
        eval_samples=eval_samples,
    )
    trainer.fit()
```

And here my questions

1: How can I set wanddb to see loss etc?
2: How can i make it use more GPU. it almost uses nothing
3: What configuration changes should I make?
4: Are there any serious issues with the outputs shown below?

![image](https://github.com/coqui-ai/TTS/assets/19240467/b232185f-5fe6-4c33-a224-40d6340f66d6)

![image](https://github.com/coqui-ai/TTS/assets/19240467/c98e38e8-4b87-435d-97e8-2f0b8fba2a29)

![image](https://github.com/coqui-ai/TTS/assets/19240467/36369511-5fe8-4271-8f1e-7eeddf637483)

**Comment by FurkanGozukara** [comment]
And here my questions

1: How can I set wanddb to see loss etc?
2: How can i make it use more GPU. it almost uses nothing. Increase number of Torch threads 
3: What configuration changes should I make?
4: Are there any serious issues with the outputs shown below?

**Reply by ehsanrs2** [reply]
OPTIMIZER_WD_ONLY_ON_WEIGHTS = True  # for multi-gpu training please make it False
to see training information  you can use tensorboard in out-path directory by this code in terminal:
`tensorboard --logdir .`

**Reply by FurkanGozukara** [reply]
thank you so much. any other changes? also how do we set wandb authorize token so training will be logged there?

and since v2 released i want to move to v2

any other changes?

https://github.com/coqui-ai/TTS/blob/dev/recipes/ljspeech/xtts_v2/train_gpt_xtts.py

**Comment by R4ZZ3** [comment]
I would also like to see loss in wandb and other metrics

**Reply by R4ZZ3** [reply]
Actually this works. Needed to just wait a bit

**Reply by FurkanGozukara** [reply]
how did you set it? wanddb token how did you set?

**Reply by R4ZZ3** [reply]
I just changed DASHBOARD_LOGGER = 'wandb' and it just worked (I had previously logged in to wandb and maybe I have it also as some env variable'

**Reply by FurkanGozukara** [reply]
that is great but you don't know how to set key for this? i would appreciate the answer

**Comment by FurkanGozukara** [comment]
> I would also like to see loss in wandb and other metrics

I am also still waiting

I written them on twitter on discord still 0 answers

**Comment by erogol** [comment]
Qs go to discussions.

**Comment by bekodata** [comment]
Hello all, also looking for a way to finetune a specific voice of choice to make it work. Please let us know.

**Comment by erogol** [comment]
1: How can I set wanddb to see loss etc?

- We don't use Wandb. There was a contribution to make it work with our [trainer](https://github.com/coqui-ai/Trainer) but I don't know if it is still compatible. You can give it a try. However, you need to figure it out yourself. There are some wandb specific [parameters](https://github.com/coqui-ai/Trainer/blob/47781f58d2714d8139dc00f57dbf64bcc14402b7/trainer/trainer.py#L65) you need to set in the config.

2: How can i make it use more GPU. it almost uses nothing. Increase number of Torch threads

- XTTS funetuning code does not support multiple GPUs currently. We don't have a timeline for it. 

3: What configuration changes should I make?

- You just start with the [recipe](https://tts.readthedocs.io/en/latest/models/xtts.html#training), update the data path and that's all.  

4: Are there any serious issues with the outputs shown below?

- They are all warnings. Don't worry about those.

**Reply by FurkanGozukara** [reply]
@erogol Thank you so much for answers

1 : @R4ZZ3 said that it is working a little bit above. he said  DASHBOARD_LOGGER = 'wandb'. any idea how to set token?
2 : I am trying to train on a single GPU : RTX 3090 TI - but it uses single torch thread - 
![image](https://github.com/coqui-ai/TTS/assets/19240467/504b3b24-fde4-4496-b767-0cce7402f622)

After fine tuning we still need to give that trained voice as reference right? thank you

//this is fine tuned model with Voice X
XTTS_CHECKPOINT = "recipes/ljspeech/xtts_v1/run/training/GPT_XTTS_LJSpeech_FT/best_model.pth" 
//this reference is Voice X
SPEAKER_REFERENCE = "LjSpeech_reference.wav"

My aim is making a tutorial

And few more questions if you could answer
in the recipe 
 max_wav_length=255995,  # ~11.6 seconds

so i should make my training wav files maximum 11.5 seconds right?
what should be minimum 3 seconds?


could you also add some description to below ones?

```
        num_loader_workers=8, // what does this do? should be left default?
        eval_split_max_size=256,  // what does this do? should be left default?
        print_step=50, // what does this do? should be left default?
        plot_step=100, // what does this do? should be left default?
        log_model_step=1000, // what does this do? should be left default?
        save_step=10000, // this is somewhat explanatory
        save_n_checkpoints=1, //  lets say I make this 5 it will save checkpoints at 10k 20k 30k 40k 50k? 
        save_checkpoints=True,
        print_eval=False, // what does this do? 
```

**Reply by erogol** [reply]
Sorry no idea about Wandb. 
Torch thread is more about CPU load. You can set it in the model config but I'd keep it one. 
Yes you need to give a reference after fine-tuning. 

I ping @Edresson to expand on fine-tuning.

**Reply by Edresson** [reply]
Yes, after fine tuning you still need to give the reference to clone the voice.

To increase the GPU VRAM usage you can increase the BATCH_SIZE on the recipe.

About the parameters that you asked for you can check the configs: https://github.com/coqui-ai/TTS/blob/6f1cba2f81de6c97e81c0b2030b631e588047968/TTS/config/shared_configs.py#L249C26-L249C39
and https://github.com/coqui-ai/Trainer/blob/47781f58d2714d8139dc00f57dbf64bcc14402b7/trainer/trainer.py#L65

**Reply by FurkanGozukara** [reply]
@erogol @Edresson 

Thank you so much. I had install error with Torch CUDA fixed it

Now uses GPU as well

Started v2 training there are literally 0 warnings amazing

CUDA speed is also amazing

I have few questions


1. how many total steps you would suggest to train a single voice? lets say 100 wav files and like 10000 steps? 
2. how can i make it generate test speech after every checkpoint file generated?
3. the batch size directly reduces number of steps necessary for 1 epoch, which makes total sense but GRAD_ACUMM_STEPS has 0 effect on it. Is that supposed to be like that? isn't GRAD_ACUMM_STEPS equal to fake batch size?
4. do you suggest float or BF16 precision? 
5. what is eval step doing? 
6. difference between epoch and eval step?
7. how many steps will eval make? 
8.  what is purpose of eval without training? 
9. Will increasing batch * grad more than 252 reduce training quality?  Like making it 4 * 84 
10. Or should we always make batch size * grad = 252? Like 4*63
11. also what does batch_group_size and eval_split_max_size do?
12. After EVAL step I get below output is that expected?
13. log_model_step generates a 2gb files under  `C:\Users\King\AppData\Local\wandb\wandb\artifacts\staging` what does it do? 
14. what does plot_step do? 

`The OrderedVocab you are attempting to save contains a hole for index 5025, your vocabulary could be corrupted !`

![image](https://github.com/coqui-ai/TTS/assets/19240467/3c10054c-66a8-478d-9db8-ea01826f610b)

@Edresson ty so much that link was super useful
